### PR TITLE
Use Map in SegmentStats

### PR DESCRIPTION
**Pull request by rjernst**
SegmentStats has an ImmutableOpenMap tracking the files that exist
within an index segment. This commit converts the member to a Map. It
also slightly tweaks the contract of the class so that the map is no
longer immutable. The reason for this is that the class as a whole is
used as an accumulator when building stats for a segment. While doing
so, the other members of SegmentStats are modified. Yet with keeping
track of files, the map is needlessly rebuilt. Instead this commit
returns an immutable view for consumers, but keeps the map internally as
mutable for adding more stats. Note that the class is already not thread
safe, so immutability is not needed here for that purpose.

relates #86239

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/engine/SegmentsStats.java [hunk 1]
```diff
@@ -144,8 +132,8 @@ public ByteSizeValue getBitsetMemory() {
         return new ByteSizeValue(bitsetMemoryInBytes);
     }
 
-    public ImmutableOpenMap<String, FileStats> getFiles() {
-        return files;
+    public Map<String, FileStats> getFiles() {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/index/engine/SegmentsStats.java:135]: This method appears to be only used in tests. Lets add a comment to that effect so nobody starts using this outright and save ourselves the unmodifiable wrapping?

File: server/src/main/java/org/elasticsearch/index/engine/SegmentsStats.java [hunk 2]
```diff
@@ -48,14 +51,7 @@ public SegmentsStats(StreamInput in) throws IOException {
...
         maxUnsafeAutoIdTimestamp = in.readLong();
-
-        final int size = in.readVInt();
-        final ImmutableOpenMap.Builder<String, FileStats> files = ImmutableOpenMap.builder(size);
-        for (int i = 0; i < size; i++) {
-            FileStats file = new FileStats(in);
-            files.put(file.getExt(), file);
-        }
-        this.files = files.build();
+        files = in.readList(FileStats::new).stream().collect(Collectors.toMap(FileStats::getExt, Function.identity()));
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/index/engine/SegmentsStats.java:54]: This one could be a little hot in some cases I believe. Let's keep the previous pattern and pre-size with `org.elasticsearch.common.util.Maps#newMapWithExpectedSize` that does without so many allocations for the list/stream etc.?

File: server/src/main/java/org/elasticsearch/index/engine/SegmentsStats.java [hunk 3]
```diff
@@ -78,18 +74,10 @@ public void addBitsetMemoryInBytes(long bitsetMemoryInBytes) {
...
-        for (Map.Entry<String, FileStats> entry : files.entrySet()) {
-            final String extension = entry.getKey();
-            if (map.containsKey(extension)) {
-                FileStats previous = map.get(extension);
-                map.put(extension, FileStats.merge(previous, entry.getValue()));
-            } else {
-                map.put(extension, entry.getValue());
-            }
+    public void addFiles(Map<String, FileStats> newFiles) {
+        for (var newEntry : newFiles.entrySet()) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/index/engine/SegmentsStats.java:78]: nice :) though NIT, could just go all the way to
```java
newFiles.forEach((key, value) -> files.merge(key, value, FileStats::merge));
```

File: server/src/main/java/org/elasticsearch/index/engine/SegmentsStats.java [hunk 4]
```diff
@@ -29,7 +32,7 @@ public class SegmentsStats implements Writeable, ToXContentFragment {
     private long versionMapMemoryInBytes;
     private long maxUnsafeAutoIdTimestamp = Long.MIN_VALUE;
     private long bitsetMemoryInBytes;
-    private ImmutableOpenMap<String, FileStats> files = ImmutableOpenMap.of();
+    private Map<String, FileStats> files = new HashMap<>();
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/index/engine/SegmentsStats.java:35]: Related to the other constructor comment. Make this `final` but assign in the constructors so we can pre-size it when reading from `StreamInput`?

**Review by original-brownbear — COMMENTED** [review]
Just a few random details.

**Comment by rjernst** [comment]
@original-brownbear I believe I addressed all your comments.

File: server/src/main/java/org/elasticsearch/index/engine/SegmentsStats.java [hunk 5]
```diff
@@ -50,12 +54,11 @@ public SegmentsStats(StreamInput in) throws IOException {
         maxUnsafeAutoIdTimestamp = in.readLong();
 
         final int size = in.readVInt();
-        final ImmutableOpenMap.Builder<String, FileStats> files = ImmutableOpenMap.builder(size);
+        files = Maps.newMapWithExpectedSize(size);
         for (int i = 0; i < size; i++) {
             FileStats file = new FileStats(in);
             files.put(file.getExt(), file);
         }
-        this.files = files.build();
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/index/engine/SegmentsStats.java:58]: Could be simplified with `this.files = in.readMapValues(FileStats::new, FileStats::getExt);`

**Review by original-brownbear — APPROVED** [review]
LGTM!

**Pull request merged**